{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Proyecto Final:\n",
    "## Reconocimiento de Entidades Nombradas\n",
    "\n",
    "El objetivo es identificar entidades nombradas en un corpus del área de genética. Las etiquetas que se buscan indican si se trata de un gen o no. Debe tomarse en cuenta que las entidades nombradas pueden constar de mas de un elemento. Por tanto, se utiliza un etiquetado BIO.\n",
    "\n",
    "Las etiquetas son como siguen:\n",
    "1. __B-tag__ indica el inicio de una entidad (de izquierda a derecha).\n",
    "2. __I-tag__ indica que la palabra pertenece a una entidad etiquetada con B-tag. En este sentido, siempre debe existir una etiqueta B-tag, pero no necesariamente una I-tag.\n",
    "3. Finalmente la etiqueta __O__ indica que no se trata de una entidad nombrada.l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cadenas Ocultas de Markov\n",
    "\n",
    "Para dar solución al problema propuesto se usara el modelo de aprendizaje automático Hidden Markov Model (HMM). Es importante establecer que para construir un HMM se necesitan llevar a cabo tres etapas:\n",
    "\n",
    "\n",
    "1. __Likelihood__: Calcular la probabilidad de las observaciones dado los parámetros del modelo $\\lambda$: \n",
    "\n",
    "    $p(O|\\lambda)$\n",
    "\n",
    "Para calcular esta probabilidad se pueden usar los algoritmos _forward_ y _backward_, ambos producen el mismo resultado.\n",
    "\n",
    "2. __Decodificacion__: Encontrar la secuencia de estados que maximice las observaciones: \n",
    "\n",
    "    $argmax_S(p(S|O))$\n",
    "\n",
    "El algoritmo de _Viterbi_ se utiliza para calcular la secuencia mas probable.\n",
    "\n",
    "3. __Aprendizaje__: Ajustar los parámetros A, B y $\\Pi$ del HMM de tal forma que maximicen la probabilidad de observar una secuencia:\n",
    "\n",
    "Con el algoritmo de _Expectation Maximization_ se ajustan las matrices de probabilidades del modelo dada una secuencia."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construccion del HMM $\\lambda$\n",
    "\n",
    "Un modelo oculto de Markov se denota mediante la letra $\\lambda$ y es una 5-tupla:\n",
    "\n",
    "$$ \\lambda = (S, O, A, B, \\Pi) $$\n",
    "\n",
    "Donde: \n",
    "\n",
    "* $S = \\{s_1, ..., s_N\\}$\n",
    "        Conjunto de estados\n",
    "* $O = \\{o_1, ..., o_T\\}$\n",
    "        Conjunto de emisiones\n",
    "* $A = \\{a_{i,j}\\} = p(q_{t+1}=S_j|q_t=S_i)$\n",
    "        Probabilidad de transiciones\n",
    "* $B = \\{b_{i,j}\\} = p(q_t=o_i| q_t=S_j)$\n",
    "        Probabilidad de emisiones\n",
    "* $\\Pi = \\{\\Pi_i\\} = p(q_1=S_i)$\n",
    "        Probabilidad de estados iniciales"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from nltk import bigrams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lectura y preprocesamiento del corpus\n",
    "\n",
    "El conjunto de estados ocultos $S$ se conformara por las etiquetas y el conjunto de simbolos de observación $O$ seran las palabras."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Final1/data_test.txt', 'r') as file:\n",
    "    raw_corpus = file.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IL-2\\tB-DNA',\n",
       " 'gene\\tI-DNA',\n",
       " 'expression\\tO',\n",
       " 'and\\tO',\n",
       " 'NF-kappa\\tB-protein',\n",
       " 'B\\tI-protein',\n",
       " 'activation\\tO',\n",
       " 'through\\tO',\n",
       " 'CD28\\tB-protein',\n",
       " 'requires\\tO']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_corpus[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Crear el conjunto de estados y observaciones junto con un diccionario\n",
    "# que almacena la secuencia de estados y emisiones como una cadena\n",
    "\n",
    "corpus = {\n",
    "    'states': [],\n",
    "    'obs': []\n",
    "}\n",
    "\n",
    "S = []\n",
    "O = []\n",
    "\n",
    "\n",
    "string_obs = ''\n",
    "string_state = ''\n",
    "\n",
    "for phrase in raw_corpus:\n",
    "    if phrase == '' and len(string_obs) > 0 and len(string_state) > 0:\n",
    "        corpus['states'].append('<BOS> ' + string_state + ' <EOS>')\n",
    "        corpus['obs'].append('<BOS> ' + string_obs + ' <EOS>')\n",
    "        string_obs = ''\n",
    "        string_state = ''\n",
    "    try:\n",
    "        obs , state = phrase.split('\\t')\n",
    "        string_obs += obs + ' ' \n",
    "        string_state += state + ' '\n",
    "        if obs not in O:\n",
    "            O.append(obs)\n",
    "        if state not in S:\n",
    "            S.append(state)\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs2indx = {}\n",
    "state2indx = {}\n",
    "indx2state = {}\n",
    "indx2obs = {}\n",
    "\n",
    "for idx, obs in enumerate(O):\n",
    "    obs2indx[obs] = idx\n",
    "    \n",
    "for idx, state in enumerate(S):\n",
    "    indx2state[idx] = state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "['<BOS>', 'IL-2', 'gene', 'expression', 'and', 'NF-kappa', 'B', 'activation', 'through', 'CD28', 'requires', 'reactive', 'oxygen', 'production', 'by', '5-lipoxygenase', '.', '<EOS>']\n"
     ]
    }
   ],
   "source": [
    "print(len(corpus['obs'][0].split()))\n",
    "print(corpus['obs'][0].split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "['<BOS>', 'B-DNA', 'I-DNA', 'O', 'O', 'B-protein', 'I-protein', 'O', 'O', 'B-protein', 'O', 'O', 'O', 'O', 'O', 'B-protein', 'O', '<EOS>']\n"
     ]
    }
   ],
   "source": [
    "print(len(corpus['states'][0].split()))\n",
    "print(corpus['states'][0].split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Numero de estados =  11\n",
      "Alfabeto de observaciones =  22053\n"
     ]
    }
   ],
   "source": [
    "len_S = len(S)\n",
    "len_O = len(O)\n",
    "\n",
    "print('Numero de estados = ', len_S)\n",
    "print('Alfabeto de observaciones = ', len_O)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtener frecuencias de bigramas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "freq_states = {}\n",
    "freq_states2states = {}\n",
    "freq_states2obs = {}\n",
    "TRUE_LABELS = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sstate, sobs in zip(corpus['states'], corpus['obs']):\n",
    "    tags = ['<BOS>', '<EOS>']\n",
    "    for si, sj in list(bigrams(sstate.split())):\n",
    "        if (si, sj) in freq_states2states:\n",
    "            freq_states2states[(si, sj)] += 1\n",
    "        else:\n",
    "            freq_states2states[(si, sj)] = 1\n",
    "\n",
    "    for si in sstate.split():\n",
    "        if si in freq_states:\n",
    "            freq_states[si] += 1\n",
    "        else:\n",
    "            freq_states[si] = 1\n",
    "            \n",
    "    for oj, si in zip(sobs.split()[1:-1], sstate.split()[1:-1]):\n",
    "        if (si, oj) not in freq_states2obs:\n",
    "            freq_states2obs[(si, oj)] = 1\n",
    "            TRUE_LABELS.append((oj, si))\n",
    "        else:\n",
    "            freq_states2obs[(si, oj)] += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creación del modelo del lenguaje\n",
    "\n",
    "El modelo del lenguaje será construido al hacer las matrices $A$, $B$ y $\\Pi$ usando el corpus de entrenamiento que ya esta etiquetado."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11, 11)\n",
      "(11, 22053)\n",
      "(11,)\n"
     ]
    }
   ],
   "source": [
    "A = np.zeros((len_S, len_S))\n",
    "B = np.zeros((len_S, len_O))\n",
    "Pi = np.zeros(len_S)\n",
    "\n",
    "N = len_S\n",
    "T = len_O\n",
    "\n",
    "print(A.shape)\n",
    "print(B.shape)\n",
    "print(Pi.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El __smoothing Laplaciano__ se usara para calcular la probabilidad condicional:\n",
    "\n",
    "$$p(x_j|x_i) =  \\frac{fr(x_i, x_j) + 1}{fr(x_i) + N}$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def smoothingLaplacian(wi, wj, L, mode):\n",
    "    if mode == 'A':\n",
    "        si, sj = wi, wj\n",
    "        try:\n",
    "            prob = (freq_states2states[(si, sj)] + 1) / (freq_states[si] + L)\n",
    "        except: \n",
    "            prob = 1 / (freq_states[wi] + L)\n",
    "    elif mode == 'B':\n",
    "        si, oj = wi, wj\n",
    "        try:\n",
    "            prob = (freq_states2obs[(si, oj)] + 1) / (freq_states[si] + L)\n",
    "        except:\n",
    "            prob = 1 / (freq_states[si] + L)\n",
    "    elif mode == 'Pi':\n",
    "        si = wi\n",
    "        try:\n",
    "            prob = (freq_states2states[('<BOS>', si)] + 1) / (freq_states['<BOS>'] + L)\n",
    "        except:\n",
    "            prob = 1 / (freq_states['<BOS>'] + L)\n",
    "    return prob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matriz de transiciones de estado\n",
    "\n",
    "Representa la probabilidad de que el siguiente estado en el tiempo _t+1_ sea $S_j$, dado que el estado actual es $S_i$:\n",
    "\n",
    "$$A_{i,j} = p(S^{t+1}_j|S^t_i)$$\n",
    "\n",
    "$$\\sum_j A_{ij} = 1$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N):\n",
    "    for j in range(N):\n",
    "        if j == N:\n",
    "            A[i, j+1] = smoothingLaplacian(S[i] , '<EOS>', N, 'A')\n",
    "        else:\n",
    "            A[i, j] = smoothingLaplacian(S[i], S[j], N, 'A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.9999999999999999,\n",
       " 0.9997465788139889,\n",
       " 0.9516157264564267,\n",
       " 0.999900826446281,\n",
       " 0.9998389499536982,\n",
       " 0.9995537042546863,\n",
       " 0.9997712978845054,\n",
       " 0.9997395154988278,\n",
       " 1.0,\n",
       " 1.0,\n",
       " 1.0]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(A.sum(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>B-DNA</th>\n",
       "      <th>I-DNA</th>\n",
       "      <th>O</th>\n",
       "      <th>B-protein</th>\n",
       "      <th>I-protein</th>\n",
       "      <th>B-cell_type</th>\n",
       "      <th>I-cell_type</th>\n",
       "      <th>B-cell_line</th>\n",
       "      <th>I-cell_line</th>\n",
       "      <th>B-RNA</th>\n",
       "      <th>I-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-DNA</th>\n",
       "      <td>0.001048</td>\n",
       "      <td>0.784345</td>\n",
       "      <td>0.213245</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000210</td>\n",
       "      <td>0.000105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-DNA</th>\n",
       "      <td>0.003548</td>\n",
       "      <td>0.525215</td>\n",
       "      <td>0.470223</td>\n",
       "      <td>0.000317</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>0.024216</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.825477</td>\n",
       "      <td>0.072979</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.017023</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.009588</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-protein</th>\n",
       "      <td>0.000793</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.559240</td>\n",
       "      <td>0.007273</td>\n",
       "      <td>0.432132</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-protein</th>\n",
       "      <td>0.000564</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.501228</td>\n",
       "      <td>0.024278</td>\n",
       "      <td>0.473326</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-cell_type</th>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.251562</td>\n",
       "      <td>0.001339</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.744719</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.000149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-cell_type</th>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.570383</td>\n",
       "      <td>0.001372</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.426529</td>\n",
       "      <td>0.000457</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-cell_line</th>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.149779</td>\n",
       "      <td>0.000781</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000781</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.846575</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-cell_line</th>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.430717</td>\n",
       "      <td>0.001218</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.007578</td>\n",
       "      <td>0.558999</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>0.000135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-RNA</th>\n",
       "      <td>0.002079</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.133056</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.856549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-RNA</th>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.534718</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.459442</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                B-DNA     I-DNA         O  B-protein  I-protein  B-cell_type  \\\n",
       "B-DNA        0.001048  0.784345  0.213245   0.000524   0.000105     0.000105   \n",
       "I-DNA        0.003548  0.525215  0.470223   0.000317   0.000063     0.000063   \n",
       "O            0.024216  0.000003  0.825477   0.072979   0.000003     0.017023   \n",
       "B-protein    0.000793  0.000033  0.559240   0.007273   0.432132     0.000132   \n",
       "I-protein    0.000564  0.000040  0.501228   0.024278   0.473326     0.000081   \n",
       "B-cell_type  0.000149  0.000149  0.251562   0.001339   0.000149     0.000893   \n",
       "I-cell_type  0.000229  0.000114  0.570383   0.001372   0.000114     0.000229   \n",
       "B-cell_line  0.000260  0.000260  0.149779   0.000781   0.000260     0.000781   \n",
       "I-cell_line  0.000135  0.000135  0.430717   0.001218   0.000135     0.000406   \n",
       "B-RNA        0.002079  0.001040  0.133056   0.001040   0.001040     0.001040   \n",
       "I-RNA        0.000649  0.000649  0.534718   0.000649   0.000649     0.000649   \n",
       "\n",
       "             I-cell_type  B-cell_line  I-cell_line     B-RNA     I-RNA  \n",
       "B-DNA           0.000105     0.000105     0.000105  0.000210  0.000105  \n",
       "I-DNA           0.000063     0.000063     0.000063  0.000063  0.000063  \n",
       "O               0.000003     0.009588     0.000003  0.002320  0.000003  \n",
       "B-protein       0.000033     0.000099     0.000033  0.000099  0.000033  \n",
       "I-protein       0.000040     0.000040     0.000040  0.000161  0.000040  \n",
       "B-cell_type     0.744719     0.000149     0.000149  0.000149  0.000149  \n",
       "I-cell_type     0.426529     0.000457     0.000114  0.000114  0.000114  \n",
       "B-cell_line     0.000260     0.000260     0.846575  0.000260  0.000260  \n",
       "I-cell_line     0.000135     0.007578     0.558999  0.000406  0.000135  \n",
       "B-RNA           0.001040     0.001040     0.001040  0.001040  0.856549  \n",
       "I-RNA           0.000649     0.000649     0.000649  0.000649  0.459442  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Aij = pd.DataFrame(A, columns=S, index=S) \n",
    "Aij"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matriz de emisiones de observaciones\n",
    "\n",
    "Esta matriz describe la probabilidad de que en el estado $S^t_i$ se emita la observación $O^t_j$:\n",
    "\n",
    "$$ B_{i,j} = p(O^t_j|S^t_i) $$\n",
    "\n",
    "$$ \\sum_j B_{i,j} = 1 $$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N):\n",
    "    for j in range(T):\n",
    "        B[i, j] = smoothingLaplacian(S[i], O[j], T, 'B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.9999999999999999,\n",
       " 1.0,\n",
       " 0.9999999999999999,\n",
       " 1.0,\n",
       " 0.9999999999999998,\n",
       " 0.9999999999999998,\n",
       " 1.0,\n",
       " 1.0,\n",
       " 0.9999999999999999,\n",
       " 1.0000000000000002,\n",
       " 1.0]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(B.sum(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IL-2</th>\n",
       "      <th>gene</th>\n",
       "      <th>expression</th>\n",
       "      <th>and</th>\n",
       "      <th>NF-kappa</th>\n",
       "      <th>B</th>\n",
       "      <th>activation</th>\n",
       "      <th>through</th>\n",
       "      <th>CD28</th>\n",
       "      <th>requires</th>\n",
       "      <th>...</th>\n",
       "      <th>80-</th>\n",
       "      <th>chromatin-assembled</th>\n",
       "      <th>-assembly</th>\n",
       "      <th>packaging</th>\n",
       "      <th>counteracted</th>\n",
       "      <th>TFE-3</th>\n",
       "      <th>Individually</th>\n",
       "      <th>HMG-88</th>\n",
       "      <th>nucleosomal</th>\n",
       "      <th>bending</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-DNA</th>\n",
       "      <td>0.005446</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.002944</td>\n",
       "      <td>0.000665</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000317</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-DNA</th>\n",
       "      <td>0.000820</td>\n",
       "      <td>0.036536</td>\n",
       "      <td>0.001163</td>\n",
       "      <td>0.008724</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>0.006133</td>\n",
       "      <td>0.000449</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.002082</td>\n",
       "      <td>0.007483</td>\n",
       "      <td>0.029831</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.005437</td>\n",
       "      <td>0.001225</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-protein</th>\n",
       "      <td>0.010766</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.019544</td>\n",
       "      <td>0.000861</td>\n",
       "      <td>0.000344</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.002467</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-protein</th>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.001557</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.008170</td>\n",
       "      <td>0.001408</td>\n",
       "      <td>0.033426</td>\n",
       "      <td>0.000853</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-cell_type</th>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.009734</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-cell_type</th>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.006301</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.006431</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-cell_line</th>\n",
       "      <td>0.000155</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.001468</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-cell_line</th>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.004213</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.003771</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-RNA</th>\n",
       "      <td>0.000869</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-RNA</th>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.002035</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.001187</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 22053 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 IL-2      gene  expression       and  NF-kappa         B  \\\n",
       "B-DNA        0.005446  0.000158    0.000158  0.000032  0.002944  0.000665   \n",
       "I-DNA        0.000820  0.036536    0.001163  0.008724  0.000608  0.006133   \n",
       "O            0.000042  0.002082    0.007483  0.029831  0.000035  0.000553   \n",
       "B-protein    0.010766  0.000096    0.000019  0.000019  0.019544  0.000861   \n",
       "I-protein    0.000299  0.001557    0.000021  0.008170  0.001408  0.033426   \n",
       "B-cell_type  0.000070  0.000035    0.000035  0.000035  0.000035  0.009734   \n",
       "I-cell_type  0.000065  0.000032    0.000065  0.006301  0.000032  0.006431   \n",
       "B-cell_line  0.000155  0.000039    0.000039  0.000039  0.000039  0.001468   \n",
       "I-cell_line  0.000136  0.000034    0.000034  0.004213  0.000034  0.003771   \n",
       "B-RNA        0.000869  0.000043    0.000043  0.000043  0.000130  0.000043   \n",
       "I-RNA        0.000085  0.000212    0.000042  0.002035  0.000042  0.001187   \n",
       "\n",
       "             activation   through      CD28  requires  ...       80-  \\\n",
       "B-DNA          0.000032  0.000032  0.000317  0.000032  ...  0.000032   \n",
       "I-DNA          0.000449  0.000026  0.000053  0.000026  ...  0.000026   \n",
       "O              0.005437  0.001225  0.000005  0.000306  ...  0.000005   \n",
       "B-protein      0.000344  0.000019  0.002467  0.000019  ...  0.000019   \n",
       "I-protein      0.000853  0.000021  0.000256  0.000021  ...  0.000021   \n",
       "B-cell_type    0.000035  0.000035  0.000035  0.000035  ...  0.000035   \n",
       "I-cell_type    0.000032  0.000032  0.000032  0.000032  ...  0.000032   \n",
       "B-cell_line    0.000039  0.000039  0.000039  0.000039  ...  0.000039   \n",
       "I-cell_line    0.000034  0.000034  0.000034  0.000034  ...  0.000034   \n",
       "B-RNA          0.000043  0.000043  0.000043  0.000043  ...  0.000043   \n",
       "I-RNA          0.000042  0.000042  0.000042  0.000042  ...  0.000042   \n",
       "\n",
       "             chromatin-assembled  -assembly  packaging  counteracted  \\\n",
       "B-DNA                   0.000063   0.000032   0.000032      0.000032   \n",
       "I-DNA                   0.000026   0.000026   0.000026      0.000026   \n",
       "O                       0.000002   0.000005   0.000005      0.000005   \n",
       "B-protein               0.000019   0.000019   0.000019      0.000019   \n",
       "I-protein               0.000021   0.000021   0.000021      0.000021   \n",
       "B-cell_type             0.000035   0.000035   0.000035      0.000035   \n",
       "I-cell_type             0.000032   0.000032   0.000032      0.000032   \n",
       "B-cell_line             0.000039   0.000039   0.000039      0.000039   \n",
       "I-cell_line             0.000034   0.000034   0.000034      0.000034   \n",
       "B-RNA                   0.000043   0.000043   0.000043      0.000043   \n",
       "I-RNA                   0.000042   0.000042   0.000042      0.000042   \n",
       "\n",
       "                TFE-3  Individually    HMG-88  nucleosomal   bending  \n",
       "B-DNA        0.000032      0.000032  0.000032     0.000063  0.000032  \n",
       "I-DNA        0.000026      0.000026  0.000026     0.000026  0.000026  \n",
       "O            0.000002      0.000005  0.000002     0.000002  0.000005  \n",
       "B-protein    0.000057      0.000019  0.000038     0.000019  0.000019  \n",
       "I-protein    0.000021      0.000021  0.000021     0.000021  0.000021  \n",
       "B-cell_type  0.000035      0.000035  0.000035     0.000035  0.000035  \n",
       "I-cell_type  0.000032      0.000032  0.000032     0.000032  0.000032  \n",
       "B-cell_line  0.000039      0.000039  0.000039     0.000039  0.000039  \n",
       "I-cell_line  0.000034      0.000034  0.000034     0.000034  0.000034  \n",
       "B-RNA        0.000043      0.000043  0.000043     0.000043  0.000043  \n",
       "I-RNA        0.000042      0.000042  0.000042     0.000042  0.000042  \n",
       "\n",
       "[11 rows x 22053 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bij = pd.DataFrame(B, columns=O, index=S)\n",
    "Bij"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Vector de probabilidades iniciales\n",
    "\n",
    "Se representa la probabilidad de que el estado $S_i$ sea el inicial:\n",
    "\n",
    "$$\\Pi_i = p(S^{t=1}_i|<BOS>)$$\n",
    "\n",
    "$$\\sum_i \\Pi_i = 1$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N):\n",
    "    Pi[i] = smoothingLaplacian(S[i], '', N, 'Pi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Pi.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;BOS&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-DNA</th>\n",
       "      <td>0.008788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-DNA</th>\n",
       "      <td>0.000054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>0.894490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-protein</th>\n",
       "      <td>0.078122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-protein</th>\n",
       "      <td>0.000054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-cell_type</th>\n",
       "      <td>0.009866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-cell_type</th>\n",
       "      <td>0.000054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-cell_line</th>\n",
       "      <td>0.005391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-cell_line</th>\n",
       "      <td>0.000054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-RNA</th>\n",
       "      <td>0.003073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I-RNA</th>\n",
       "      <td>0.000054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                <BOS>\n",
       "B-DNA        0.008788\n",
       "I-DNA        0.000054\n",
       "O            0.894490\n",
       "B-protein    0.078122\n",
       "I-protein    0.000054\n",
       "B-cell_type  0.009866\n",
       "I-cell_type  0.000054\n",
       "B-cell_line  0.005391\n",
       "I-cell_line  0.000054\n",
       "B-RNA        0.003073\n",
       "I-RNA        0.000054"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Pii = pd.DataFrame(Pi, columns=['<BOS>'], index=S)\n",
    "Pii"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creación del modelo de Markov\n",
    "\n",
    "Para encontrar la probabilidad de las observaciones dado el modelo $\\lambda$ se aplican los algoritmos forward y backward.\n",
    "\n",
    "$$p(o_1,\\dots, o_T| \\lambda)$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Algoritmo Forward\n",
    "\n",
    "Se define la variable forward como:\n",
    "\n",
    "$\\alpha_t(j) = p(o_1, ..., o_t, q_t=S_j|\\lambda)$\n",
    "\n",
    "Para encontrar en cada iteración la probabilidad conjunta de las observaciones se define:\n",
    "\n",
    "$\\alpha_1(j) = \\Pi_j b_j(o_1)$\n",
    "\n",
    "$\\alpha_{t+1}(j) = \\sum_{i=1}^N \\alpha_{t}(i) A_{i,j} b_j(o_t)$\n",
    "\n",
    "En forma matricial se puede escribir como:\n",
    "\n",
    "$\\alpha_{t+1} = b(o_t) \\odot ( A^T \\bullet \\alpha_t)$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Algoritmo Backward\n",
    "\n",
    "Para encontrar en cada iteracion la probabilidad conjunta de las observaciones se define:\n",
    "\n",
    "$\\beta_T(j) = 1 $\n",
    "\n",
    "$\\beta_t(j) = \\sum_i A_{j,i} b_i(O_{t+1}) \\beta_{t+1}(i) $"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Decodificacion\n",
    "\n",
    "Se utiliza el algoritmo de Viterbi para encontrar la secuencia de estados que maximiza la probabilidad de emitir las observaciones dadas.\n",
    "\n",
    "Se define la variable $\\delta$:\n",
    "\n",
    "$\\delta_0(i) = \\Pi_i b_i(O_0)$\n",
    "\n",
    "$\\delta_t(i) = max(\\delta_{t-1}(j)a_{ji}b_i(O_t))$\n",
    "\n",
    "En cada iteración del tiempo, se selecciona el estado que tenga la mayor probabilidad y se guarda la posición en una variable llamada _backpointer_ con un _argmax_ del mismo argumento de la variable $\\delta$.\n",
    "\n",
    "Si se comparan el algoritmo de Viterbi y el Forward, son basicamente el mismo algoritmo con la diferencia de que en Forward se toma la suma de las probabilidades de cada posible camino y en Viterbi se toma el camino con mayor probabilidad."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Aprendizaje \n",
    "\n",
    "Los parámetros del HMM se ajustan mediante la aplicación del algoritmo Forward-Backward que consiste de dos etapas:\n",
    "\n",
    "1. Expectation: Calcular los valores esperados de transiciones de estados.\n",
    "\n",
    "$\\gamma_t(i) = \\frac{\\alpha_t(i)\\beta_t(i)}{p(O|\\lambda)} = \\frac{\\alpha_t(i)\\beta_t(i)}{\\alpha_T}$\n",
    "\n",
    "$\\xi_t(i,j) = \\frac{\\alpha_t(i)a_{ij}b_j(O_{t+1}\\beta_{t+1}(j))}{\\alpha_T}$\n",
    "\n",
    "2. Maximization: Actualizar los parámetros del modelo.\n",
    "\n",
    "$a_{ij} = \\frac{\\sum_t\\xi_t(i,j)}{\\sum_t\\sum_j\\xi_t(i,j)}$\n",
    "\n",
    "$b_i(k) = \\frac{\\sum_t \\begin{cases} 0 & O_t \\neq V_k \\\\ \\gamma_t(i) & O_t = V_k \\end{cases} }{\\sum_t\\gamma_t(k)}$\n",
    "\n",
    "$\\Pi_i = \\alpha_0(i)$\n",
    "\n",
    "Este proceso se realiza hasta llegar a cierto numero de iteraciones o que se cumpla algún criterio."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HMM():\n",
    "    def __init__(self, S, O, A, B, Pi):\n",
    "        self.S = S\n",
    "        self.O = O\n",
    "        self.A = A\n",
    "        self.B = B\n",
    "        self.Pi = Pi\n",
    "        self.N = len(S)\n",
    "        self.T = len(O)\n",
    "    \n",
    "    def forward(self, sequence):\n",
    "        words = sequence.split()\n",
    "        alpha = np.zeros((self.N, len(words)))\n",
    "        alpha[:, 0] = self.Pi * Bij[words[0]]\n",
    "        t = np.arange(len(words))\n",
    "        for word, ti in zip(words[1:], t[1:]):\n",
    "            alpha[:, ti] = np.dot(self.A.T, alpha[:, ti-1]) * self.B[:, obs2indx[word]]\n",
    "        return alpha[:,-1].sum(), alpha\n",
    "            \n",
    "    def backward(self, sequence):\n",
    "        words = sequence.split()\n",
    "        beta = np.zeros((self.N, len(words)))\n",
    "        beta[:,-1] = np.ones(self.N)\n",
    "        for t in range(len(words)-1, 0, -1):\n",
    "            for j in range(self.N):\n",
    "                beta[j, t-1] = (self.A[j,:] * self.B[:, obs2indx[words[t]]] * beta[:, t]).sum()\n",
    "        return (self.Pi * beta[:, 0] * self.B[:, obs2indx[words[0]]]).sum(), beta\n",
    "    \n",
    "    def viterbi(self, seq):\n",
    "        words = seq.split()\n",
    "        delta = np.zeros((self.N, len(words)))\n",
    "        delta[:,0] = self.Pi * self.B[:, obs2indx[words[0]]]\n",
    "        bckptr = np.zeros((self.N, len(words)))\n",
    "        path = []\n",
    "        for t in range(len(words)-1):\n",
    "            for j in range(self.N):\n",
    "                try:\n",
    "                    z = self.A[:, j] * delta[:, t] * self.B[:, obs2indx[words[t+1]]]\n",
    "                except: \n",
    "                    z = self.A[:, j] * delta[:, t] * 1\n",
    "                delta[j, t+1] = np.max(z)\n",
    "                bckptr[j, t+1] = np.argmax(z)\n",
    "        prob = np.max(delta[:,-1], axis=0)\n",
    "        for t in range(len(words)):\n",
    "            z = np.max(delta[:, t])\n",
    "            zidx = np.argmax(delta[:, t])\n",
    "            path.append(zidx)\n",
    "        return path\n",
    "    \n",
    "    def forward_backward(self, seq, steps):\n",
    "        words = seq.split()\n",
    "        alphaT, alpha = self.forward(seq)\n",
    "        beta0, beta = self.backward(seq) \n",
    "        gamma = np.zeros((self.N, len(words)))\n",
    "        E = np.zeros((len(words), self.N, self.N))\n",
    "        for i in range(steps):\n",
    "            # E-step\n",
    "            for t in range(len(words)):\n",
    "                for j in range(self.N):\n",
    "                    gamma[j, t] = (alpha[j, t]*beta[j, t])/alphaT\n",
    "            \n",
    "            for t in range(len(words)-1):\n",
    "                for i in range(self.N):\n",
    "                    for j in range(self.N):\n",
    "                        E[t, i, j] = (alpha[i, t]*self.A[i,j]*self.B[j, obs2indx[words[t+1]]]*beta[j, t+1])/alphaT\n",
    "            # M-step\n",
    "            for i in range(self.N):\n",
    "                for j in range(self.N):\n",
    "                    self.A[i, j] = np.sum(E[:-1, i, j])/np.sum([np.sum(E[t, i, :]) for t in range(len(words)-1)])\n",
    "            \n",
    "            for j in range(self.N):\n",
    "                for k in range(self.T):\n",
    "                    num = 0\n",
    "                    for t in range(len(words)):\n",
    "                        if obs2indx[words[t]] == k:\n",
    "                            num += gamma[j, t]\n",
    "                    #self.B[j, k] = num / np.sum(gamma[j, :])\n",
    "            self.Pi = gamma[:, 0]\n",
    "        #return gamma, E\n",
    "        \n",
    "    def score(self, predicted_labels, true_labels):\n",
    "        sc = 0\n",
    "        N = len(predicted_labels)\n",
    "        for pred_pair, true_pair in zip(predicted_labels, true_labels):\n",
    "            if pred_pair[1] == true_pair[1]:\n",
    "                sc += 1\n",
    "        return sc/N"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predecir los estados o tags de una secuencia dado un modelo de markov\n",
    "def predictTags(hmm, seq, RTG=False):\n",
    "    words = seq.split()\n",
    "    path = hmm.viterbi(seq)\n",
    "    if RTG:\n",
    "        return [(word, indx2state[itag]) for itag, word in zip(path, words)]\n",
    "    for word, p in zip(words, path):\n",
    "        print(f'{word}\\t{indx2state[p]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Entrenamiento del HMM\n",
    "\n",
    "Se muestra el entrenamiento del HMM con una oracion en particular del corpus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "hmm = HMM(S, O, A, B, Pi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha = 3.9589842025343836e-47 \t beta = 3.958984202534385e-47\n",
      "alpha = 2.5586096125373176e-75 \t beta = 2.558609612537318e-75\n",
      "alpha = 2.6646760772142282e-114 \t beta = 2.6646760772142282e-114\n",
      "alpha = 4.713049060145071e-71 \t beta = 4.713049060145066e-71\n",
      "alpha = 3.437866893431722e-60 \t beta = 3.4378668934317234e-60\n",
      "alpha = 1.3337146861198805e-58 \t beta = 1.3337146861198798e-58\n",
      "alpha = 1.0793322882318052e-55 \t beta = 1.079332288231806e-55\n",
      "alpha = 1.940121066137957e-77 \t beta = 1.94012106613796e-77\n",
      "alpha = 4.1462103665387673e-57 \t beta = 4.1462103665387684e-57\n",
      "alpha = 1.9760470421394936e-66 \t beta = 1.9760470421394952e-66\n"
     ]
    }
   ],
   "source": [
    "# Ejecucion del algoritmo forward y backward para el calculo de la probabilidad \n",
    "# de las observaciones dado el modelo de markov definido para las primeras\n",
    "# 10 oraciones. El valor de alpha y beta debe ser el mismo como se verifica:\n",
    "for seq in corpus['obs'][:10]:\n",
    "    seq = seq.strip('<[B|E]OS>')\n",
    "    alphaT , _ = hmm.forward(seq)\n",
    "    beta0, _ = hmm.backward(seq)\n",
    "    print('alpha = {} \\t beta = {}'.format(alphaT, beta0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Secuencia de observaciones: \n",
      " IL-2 gene expression and NF-kappa B activation through CD28 requires reactive oxygen production by 5-lipoxygenase .  \n",
      "\n",
      "Secuencia de estados:\n",
      "[3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]\n"
     ]
    }
   ],
   "source": [
    "# Algoritmo de Viterbi para la primera oracion\n",
    "# Los numeros de la lista representan el indice de los estados o tags\n",
    "\n",
    "seq = corpus['obs'][0].strip('<[B|E]OS>')\n",
    "\n",
    "print('Secuencia de observaciones: ')\n",
    "print(seq)\n",
    "\n",
    "print('\\nSecuencia de estados:')\n",
    "print(hmm.viterbi(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prediccion:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('IL-2', 'B-protein'),\n",
       " ('gene', 'O'),\n",
       " ('expression', 'O'),\n",
       " ('and', 'O'),\n",
       " ('NF-kappa', 'O'),\n",
       " ('B', 'O'),\n",
       " ('activation', 'O'),\n",
       " ('through', 'O'),\n",
       " ('CD28', 'O'),\n",
       " ('requires', 'O'),\n",
       " ('reactive', 'O'),\n",
       " ('oxygen', 'O'),\n",
       " ('production', 'O'),\n",
       " ('by', 'O'),\n",
       " ('5-lipoxygenase', 'O'),\n",
       " ('.', 'O')]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Prediccion:')\n",
    "predictTags(hmm, seq, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('IL-2', 'B-DNA'),\n",
       " ('gene', 'I-DNA'),\n",
       " ('expression', 'O'),\n",
       " ('and', 'O'),\n",
       " ('NF-kappa', 'B-protein'),\n",
       " ('B', 'I-protein'),\n",
       " ('activation', 'O'),\n",
       " ('through', 'O'),\n",
       " ('CD28', 'B-protein'),\n",
       " ('requires', 'O'),\n",
       " ('reactive', 'O'),\n",
       " ('oxygen', 'O'),\n",
       " ('production', 'O'),\n",
       " ('by', 'O'),\n",
       " ('5-lipoxygenase', 'B-protein'),\n",
       " ('.', 'O')]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Se imprime el verdadero etiquetado de la oracion\n",
    "TRUE_LABELS[:16]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Se entrena el HMM para la primer oracion\n",
    "hmm.forward_backward(seq, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nueva prediccion:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('IL-2', 'B-protein'),\n",
       " ('gene', 'O'),\n",
       " ('expression', 'B-protein'),\n",
       " ('and', 'B-protein'),\n",
       " ('NF-kappa', 'O'),\n",
       " ('B', 'O'),\n",
       " ('activation', 'B-protein'),\n",
       " ('through', 'O'),\n",
       " ('CD28', 'O'),\n",
       " ('requires', 'B-protein'),\n",
       " ('reactive', 'O'),\n",
       " ('oxygen', 'B-protein'),\n",
       " ('production', 'B-protein'),\n",
       " ('by', 'B-protein'),\n",
       " ('5-lipoxygenase', 'O'),\n",
       " ('.', 'B-protein')]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Se realiza una nueva prediccion despues del ajuste de parametros\n",
    "print('Nueva prediccion:')\n",
    "predictTags(hmm, seq, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calcular el score del hmm \n",
    "\n",
    "TRUE_LABELS[:5]\n",
    "PREDICTED_LABELS = []\n",
    "\n",
    "for seq in corpus['obs']:\n",
    "    seq = seq.strip('<[B|E]OS>')\n",
    "    PREDICTED_LABELS += predictTags(hmm, seq, True)\n",
    "    \n",
    "#SCORE = hmm.score(TRUE_LABELS, PREDICTED_LABELS)\n",
    "#print(SCORE*100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicción de estados (datos de prueba)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open (\"Final1/data_eval.txt\", \"r\") as file:\n",
    "    corpus_test = file.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HIV-1 envelope glycoproteins induce activation of activated protein-1 in CD4+ T cells [published erratum appears in J Biol Chem 1995 Dec 1; 270 (48): 29038]  Activation of CD4 positive T cells is a primary requirement for human immunodeficiency virus (HIV) entry, efficient HIV replication, and progression to AIDS, Utilizing CD4 positive T cell lines and purified T cells from normal individuals, we have demonstrated that native envelope glycoproteins of HIV, gp 160, can induce activation of transcription factor, activated protein-1 (AP-1).',\n",
       " 'The stimulatory effects of gp160 are mediated through the CD4 molecule, since treatment of gp160 with soluble CD4-IgG abrogates its activity, and CD4 negative T cell lines fail to be stimulated with gp160.',\n",
       " 'Immunoprecipitation of the gp 160-induced nuclear extracts with polyclonal antibodies to Fos and Jun proteins indicates that AP-1 complex is comprised of members of these family of proteins.',\n",
       " 'The gp160-induced AP-1 complex is dependent upon protein tyrosine phosphorylation and is protein synthesis-independent.',\n",
       " 'This stimulation can also be abolished by inhibitors of protein kinase C, but it is unaffected by calcium channel blocker or cyclosporine A.',\n",
       " 'This gp160 treatment adversely affects the functional capabilities of T cells: pre-treatment of CD4+ T cells with gp160 for 4 h at 37 degrees C inhibited anti-CD3 -induced interleukin-2 secretion.',\n",
       " 'Effects similar to gp160 were seen with anti-CD4 mAb.',\n",
       " 'The aberrant activation of AP-1 by gp160 in CD4 positive T cells could result in up-regulation of cytokines containing AP-1 sites, e.g. interleukin-3 and granulocyte macrophage colony-stimulating factor, and concurrently lead to T cell unresponsiveness by inhibiting interleukin-2 secretion.',\n",
       " '.']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "corpus_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predecir los tags de cada palabra de una secuencia \n",
    "ctags = []\n",
    "for seq in corpus_test:\n",
    "    ctags.append(predictTags(hmm, seq, RTG=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('HIV-1', 'B-DNA'),\n",
       " ('envelope', 'I-DNA'),\n",
       " ('glycoproteins', 'O'),\n",
       " ('induce', 'B-protein'),\n",
       " ('activation', 'O'),\n",
       " ('of', 'B-protein'),\n",
       " ('activated', 'O'),\n",
       " ('protein-1', 'O'),\n",
       " ('in', 'B-protein'),\n",
       " ('CD4+', 'O'),\n",
       " ('T', 'B-protein'),\n",
       " ('cells', 'B-protein'),\n",
       " ('[published', 'O'),\n",
       " ('erratum', 'B-protein'),\n",
       " ('appears', 'B-protein'),\n",
       " ('in', 'B-protein'),\n",
       " ('J', 'O'),\n",
       " ('Biol', 'B-protein'),\n",
       " ('Chem', 'O'),\n",
       " ('1995', 'B-protein'),\n",
       " ('Dec', 'O'),\n",
       " ('1;', 'B-protein'),\n",
       " ('270', 'O'),\n",
       " ('(48):', 'B-protein'),\n",
       " ('29038]', 'O'),\n",
       " ('Activation', 'B-protein'),\n",
       " ('of', 'B-protein'),\n",
       " ('CD4', 'O'),\n",
       " ('positive', 'B-protein'),\n",
       " ('T', 'O'),\n",
       " ('cells', 'B-protein'),\n",
       " ('is', 'B-protein'),\n",
       " ('a', 'B-protein'),\n",
       " ('primary', 'O'),\n",
       " ('requirement', 'B-protein'),\n",
       " ('for', 'B-protein'),\n",
       " ('human', 'O'),\n",
       " ('immunodeficiency', 'O'),\n",
       " ('virus', 'B-protein'),\n",
       " ('(HIV)', 'O'),\n",
       " ('entry,', 'B-protein'),\n",
       " ('efficient', 'B-protein'),\n",
       " ('HIV', 'O'),\n",
       " ('replication,', 'B-protein'),\n",
       " ('and', 'B-protein'),\n",
       " ('progression', 'B-protein'),\n",
       " ('to', 'B-protein'),\n",
       " ('AIDS,', 'O'),\n",
       " ('Utilizing', 'O'),\n",
       " ('CD4', 'O'),\n",
       " ('positive', 'B-protein'),\n",
       " ('T', 'O'),\n",
       " ('cell', 'B-protein'),\n",
       " ('lines', 'B-protein'),\n",
       " ('and', 'B-protein'),\n",
       " ('purified', 'O'),\n",
       " ('T', 'B-protein'),\n",
       " ('cells', 'B-protein'),\n",
       " ('from', 'B-protein'),\n",
       " ('normal', 'B-protein'),\n",
       " ('individuals,', 'B-protein'),\n",
       " ('we', 'B-protein'),\n",
       " ('have', 'O'),\n",
       " ('demonstrated', 'B-protein'),\n",
       " ('that', 'B-protein'),\n",
       " ('native', 'O'),\n",
       " ('envelope', 'O'),\n",
       " ('glycoproteins', 'O'),\n",
       " ('of', 'B-protein'),\n",
       " ('HIV,', 'O'),\n",
       " ('gp', 'B-protein'),\n",
       " ('160,', 'O'),\n",
       " ('can', 'B-protein'),\n",
       " ('induce', 'O'),\n",
       " ('activation', 'B-protein'),\n",
       " ('of', 'B-protein'),\n",
       " ('transcription', 'O'),\n",
       " ('factor,', 'B-protein'),\n",
       " ('activated', 'B-protein'),\n",
       " ('protein-1', 'O'),\n",
       " ('(AP-1).', 'B-protein')]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Predicciones de la primera secuencia\n",
    "ctags[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Referencias\n",
    "\n",
    "Para construir el HMM se consultaron las siguientes fuentes:\n",
    "1. https://dev.widemeadows.de/2014/09/01/hidden-markov-model-training-using-the-baum-welch-algorithm/\n",
    "2. https://medium.com/@kangeugine/hidden-markov-model-7681c22f5b9\n",
    "3. https://web.stanford.edu/~jurafsky/slp3/ed3book.pdf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
